File: src/main/providers/mock.js
'use strict';

/**
 * Offline provider: a seeded random walk per symbol. Needs no network access
 * and no API key, so the widget is fully usable out of the box.
 */

const walks = new Map();

function seedFrom(symbol) {
  let h = 2166136261;
  for (let i = 0; i < symbol.length; i++) {
    h ^= symbol.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return (h >>> 0) / 0xffffffff;
}

function walkFor(symbol) {
  let walk = walks.get(symbol);
  if (!walk) {
    const seed = seedFrom(symbol);
    const base = 10 + seed * 290; // 10 .. 300
    walk = {
      prevClose: Math.round(base * 100) / 100,
      price: Math.round(base * 100) / 100,
      volatility: 0.002 + seed * 0.006,
    };
    walks.set(symbol, walk);
  }
  return walk;
}

async function fetchQuotes(symbols) {
  return symbols.map((symbol) => {
    const walk = walkFor(symbol);
    const drift = (Math.random() - 0.5) * 2 * walk.volatility;
    walk.price = Math.max(0.01, Math.round(walk.price * (1 + drift) * 100) / 100);
    const change = walk.price - walk.prevClose;
    return {
      symbol,
      name: symbol,
      price: walk.price,
      prevClose: walk.prevClose,
      change,
      changePercent: walk.prevClose ? (change / walk.prevClose) * 100 : 0,
      currency: '',
      time: Date.now(),
      error: null,
    };
  });
}

module.exports = {
  id: 'mock',
  label: '模拟数据（离线，无需网络）',
  placeholder: 'DEMO, TEST, ANY-SYMBOL',
  fetchQuotes,
};

File: stockwidget/providers/mock.py
"""离线模拟行情：按代码播种的随机游走，不联网、不需要密钥。"""

from __future__ import annotations

import random
import zlib

from .base import Quote


class MockProvider:
    id = "mock"
    label = "模拟数据（离线，无需网络）"
    placeholder = "任意代码，如 600519、DEMO"

    def __init__(self) -> None:
        self._walks: dict[str, dict[str, float]] = {}

    def _walk(self, symbol: str) -> dict[str, float]:
        walk = self._walks.get(symbol)
        if walk is None:
            seed = zlib.crc32(symbol.encode("utf-8")) / 0xFFFFFFFF
            base = round(10 + seed * 290, 2)
            walk = {"prev_close": base, "price": base, "volatility": 0.002 + seed * 0.006}
            self._walks[symbol] = walk
        return walk

    def fetch(self, symbols: list[str]) -> list[Quote]:
        quotes = []
        for symbol in symbols:
            walk = self._walk(symbol)
            drift = (random.random() - 0.5) * 2 * walk["volatility"]
            walk["price"] = max(0.01, round(walk["price"] * (1 + drift), 2))
            quotes.append(
                Quote.from_prices(symbol, symbol, walk["price"], walk["prev_close"])
            )
        return quotes
